uamqp/authentication/cbs_auth_async.py

```diff
@@ -221,3 +235,56 @@ class JWTTokenAsync(JWTTokenAuth, CBSAsyncAuthMixin):
      Default is 'UTF-8'.
     :type encoding: str
     """
+
+    def __init__(self, audience, uri,
+                 get_token,
+                 expires_in=datetime.timedelta(seconds=constants.AUTH_EXPIRATION_SECS),
+                 expires_at=None,
+                 port=constants.DEFAULT_AMQPS_PORT,
+                 timeout=10,
+                 retry_policy=TokenRetryPolicy(),
+                 verify=None,
+                 token_type=b"jwt",
+                 http_proxy=None,
+                 transport_type=TransportType.Amqp,
+                 encoding='UTF-8'):  # pylint: disable=no-member
+        self._retry_policy = retry_policy
+        self._encoding = encoding
+        self.uri = uri
+        parsed = compat.urlparse(uri)  # pylint: disable=no-member
+
+        self.cert_file = verify
+        self.hostname = parsed.hostname.encode(self._encoding)
+
+        import functools
+        if isinstance(get_token, functools.partial):  # get_token is a partial function
+            if not asyncio.iscoroutinefunction(get_token.func):
+                raise ValueError("get_token must be a coroutine function")
+        elif callable(get_token):  # get_token is a function
+            if not asyncio.iscoroutinefunction(get_token):
+                raise ValueError("get_token must be a coroutine function")
+        else:
+            raise ValueError("get_token must be a coroutine function")
+
+        self.get_token = get_token
+        self.audience = self._encode(audience)
+        self.token_type = self._encode(token_type)
+        if not expires_at and not expires_in:
+            raise ValueError("Must specify either 'expires_at' or 'expires_in'.")
+        if not expires_at:
+            self.expires_in = expires_in
+            self.expires_at = time.time() + expires_in.seconds
+        else:
+            self.expires_at = expires_at
+            expires_in = expires_at - time.time()
+            if expires_in < 1:
+                raise ValueError("Token has already expired.")
+            self.expires_in = datetime.timedelta(seconds=expires_in)
```

Comment:
This section of code is duplicated 4 times between all 4 token type - we should move it to a method on the CBSAuthMixin:
```python
class CBSAuthMixin:
    def _set_expiry(self, expires_at, expires_in):
        if not expires_at and not expires_in:
            raise ValueError("Must specify either 'expires_at' or 'expires_in'.")
        if not expires_at:
            expires_at = time.time() + expires_in.seconds
        else:
            expires_in_seconds = expires_at - time.time()
        if expires_in_seconds < 1:
            raise ValueError("Token has already expired.")
            expires_in = datetime.timedelta(seconds=expires_in_seconds)
        return expires_at, expires_in

Token(CBSAuthMixin):
    def __init__(self, ...):
        ....
        self.expires_at, self.expires_in = self._set_expiry(expires_at, expires_in)
        ...
```